### PR TITLE
Account for KVS collisions in NVM identifiers

### PR DESCRIPTION
**Pull request by doru91**
#### Problem
* Old implementation was not taking into account collisions.

#### Change overview
* The new implementation keeps an unordered hash having as keys the matter keys and as values the key ids used as NVM identifiers. Used key ids are saved in a list which is updated at each Put/Delete operation.

#### Testing
* manual testing

### CODE REVIEW

File: src/platform/K32W/KeyValueStoreManagerImpl.cpp [hunk 1]
```diff
@@ -39,61 +42,89 @@ namespace PersistedStorage {
 
 KeyValueStoreManagerImpl KeyValueStoreManagerImpl::sInstance;
 
+/* hashmap having:
+ * 	- the matter key value as key;
+ * 	- internal PDM identifier as value;
+ */
+std::unordered_map<std::string, uint8_t> kvs;
```
Comment (kghost) [inline src/platform/K32W/KeyValueStoreManagerImpl.cpp:49]: We should avoid using stl containers which allocate memory in heap.

Comment (andy31415) [reply src/platform/K32W/KeyValueStoreManagerImpl.cpp:49]: What is the RAM availability of K32W?

Generally I would be ok with stl on large platforms (e.g. linux or darwin) in platform code. However here I agree that an embedded device likely should not use STL that may allocate heap.

Comment (andy31415) [reply src/platform/K32W/KeyValueStoreManagerImpl.cpp:49]: Also separate code review comments:

name of 'kvs' does not seem to indicate global. use gKvsKeyIdMap or something similar.
Where is this variable persisted? If device reboots,  how is this map and list below initialized?

**Review by andy31415 — CHANGES_REQUESTED** [review]
Please update the summary with:
  - if using stl with heap usage, why this is ok (is this temporary or ok for long term?)
  - how are the new global variables persisted across reboot? What makes the method work consistently?

**Review by woody-apple — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

File: src/platform/K32W/KeyValueStoreManagerImpl.cpp (hunk 1 above)
Comment (tcarmelveilleux) [reply src/platform/K32W/KeyValueStoreManagerImpl.cpp:49]: That is not a problem if it's in platform code. They use STL in lots of other things already on K32W

**Review by msandstedt — CHANGES_REQUESTED** [review]
>     * how are the new global variables persisted across reboot? What makes the method work consistently?

++ What @andreilitvin said.

Three things:

* kvs and kvs_ids have to be recreated at boot, but aren't
* it is impossible to recreate kvs at boot though because the string keys can't be recovered from the uint8 kv id hash
* overwriting a matching key with PUT is no longer possible; put with equal key isn't a key collision; this should be legal

I believe a similar key collision problem may have been solved here: https://github.com/project-chip/connectedhomeip/pull/7472.  Suggest using that as a possible guide to a workable solution.

**Comment by tcarmelveilleux** [comment]
@msandstedt 

> 
> I believe a similar key collision problem may have been solved here: #7472. Suggest using that as a possible guide to a workable solution.

I think the approach of #7472 is not going to work, because it does static string -> key ID mapping. That implies *no dynamic keys*. For instance, to have working storage for multiple fabrics, it would means storing ALL fabrics in 1 key with a single name, rather than being able to store one key per fabric, dynamically "generated" (e.g. `fabric1`, `fabric2`, etc). So yes, using #7472 is a very quick "short term" solution, but won't scale to needing to store a potentially dynamic amount of information, such as ACL entries, multiple fabric data and/or certs split by fabric, etc.

Overall, we are starting to arrive at a new choke point in storage examples for platforms that do not arbitrarily support key mapping.

**Comment by msandstedt** [comment]
> @msandstedt
> 
> > I believe a similar key collision problem may have been solved here: #7472. Suggest using that as a possible guide to a workable solution.
> 
> I think the approach of #7472 is not going to work, because it does static string -> key ID mapping. That implies _no dynamic keys_. For instance, to have working storage for multiple fabrics, it would means storing ALL fabrics in 1 key with a single name, rather than being able to store one key per fabric, dynamically "generated" (e.g. `fabric1`, `fabric2`, etc). So yes, using #7472 is a very quick "short term" solution, but won't scale to needing to store a potentially dynamic amount of information, such as ACL entries, multiple fabric data and/or certs split by fabric, etc.
> 
> Overall, we are starting to arrive at a new choke point in storage examples for platforms that do not arbitrarily support key mapping.

@doru91,

@tcarmelveilleux and I discussed offline and in fact #7472 does work (albeit only for keys <= 32 chars long).  It's using two separate non-volatile spaces: one for keys and one for values.  This is a variant of a table of contents.

The larger story is that we are dissatisfied with the sdk's current kvstore interface.  However, for the purposes here of this PR (#9489), you just need to find some way to persist the table of contents.  #7472 shows one possible approach.

File: src/platform/K32W/KeyValueStoreManagerImpl.cpp (hunk 1 above)
Comment (doru91) [reply src/platform/K32W/KeyValueStoreManagerImpl.cpp]: @andy31415 , K32W has 152K of SRAM. Let's keep stl for the moment and revise it later when the KVS API will change (maybe strings won't be used as keys). I also refactored the names of the global variables.

**Comment by doru91** [comment]
> Please update the summary with:
> 
> * if using stl with heap usage, why this is ok (is this temporary or ok for long term?)

Using stl is not ok for the long term, I expect that the KVS API will change to use numbers instead of strings as keys.

> * how are the new global variables persisted across reboot? What makes the method work consistently?
I modified the implementation to add a method for restoring from flash in case of restart.

File: src/platform/K32W/KeyValueStoreManagerImpl.cpp [hunk 2]
```diff
@@ -24,76 +24,189 @@
...
 
+#include <unordered_map>
+
 namespace chip {
 namespace DeviceLayer {
 namespace PersistedStorage {
 
 /* TODO: adjust this value */
-#define MAX_NO_OF_KEYS 255
+#define MAX_NO_OF_KEYS 20
```
Comment (andy31415) [inline src/platform/K32W/KeyValueStoreManagerImpl.cpp:41]: this is C++, use  `constexpr size_t kMaxNumberOfKeys = 20;`

File: src/platform/K32W/KeyValueStoreManagerImpl.cpp [hunk 3]
```diff
@@ -24,76 +24,189 @@
...
+ * 	- the matter key string as key;
+ * 	- internal PDM identifier as value;
+ */
+std::unordered_map<std::string, uint8_t> g_kvs_map;
+
+/* list containing used PDM identifiers */
+std::list<uint8_t> g_key_ids_list;
+
+/* max no of bytes for a key */
+#define MAX_KEY_VALUE          255
```
Comment (andy31415) [inline src/platform/K32W/KeyValueStoreManagerImpl.cpp]: same: use C++ constants. Ideally group constants in one place.

Use the name of `kMaxKeyValueBytes` then the comment is not needed.

File: src/platform/K32W/KeyValueStoreManagerImpl.cpp [hunk 4]
```diff
@@ -24,76 +24,189 @@
...
+
+/* used to check if we need to restore values from flash (e.g.: reset) */
+static bool g_restored_from_flash = FALSE;
```
Comment (andy31415) [inline src/platform/K32W/KeyValueStoreManagerImpl.cpp:58]: s/FALSE/false/

do this throughout.

File: src/platform/K32W/KeyValueStoreManagerImpl.cpp [hunk 5]
```diff
@@ -24,76 +24,189 @@
...
 
-CHIP_ERROR KeyValueStoreManagerImpl::_Get(const char * key, void * value, size_t value_size, size_t * read_bytes_size,
-                                          size_t offset_bytes)
+/* hashmap having:
+ * 	- the matter key string as key;
+ * 	- internal PDM identifier as value;
+ */
+std::unordered_map<std::string, uint8_t> g_kvs_map;
+
+/* list containing used PDM identifiers */
```
Comment (andy31415) [inline src/platform/K32W/KeyValueStoreManagerImpl.cpp:51]: should this be a set?

File: src/platform/K32W/KeyValueStoreManagerImpl.cpp [hunk 6]
```diff
@@ -24,76 +24,189 @@
...
-    ChipLogProgress(DeviceLayer, "KVS, get key id:: %i", key_id);
+    for (key_id = 0; key_id < MAX_NO_OF_KEYS; key_id++)
+    {
+        /* key was saved as string in flash (key_string_id) using (key_id + MAX_NO_OF_KEYS) as PDM key
+         * value corresponding to key_string_id was saved in flash using key_id as PDM key
+         */
 
-    err = chip::DeviceLayer::Internal::K32WConfig::ReadConfigValueBin(
-        chip::DeviceLayer::Internal::K32WConfigKey(pdm_id_kvs, key_id), (uint8_t *) value, value_size, read_bytes);
+        err = chip::DeviceLayer::Internal::K32WConfig::ReadConfigValueStr(
```
Comment (andy31415) [inline src/platform/K32W/KeyValueStoreManagerImpl.cpp:80]: this seems racy: if the last read is success, then this method will return succes. if the last is NOT_FOUND, the method returns NOT_FOUND.

Either consistently clear the error or return error on first failure. It sounds like the logic should be:

- if NOT_FOUND then set error to NO_ERROR and continue
- on any error except NO_ERROR or NOT_FOUND, return an error

File: src/platform/K32W/KeyValueStoreManagerImpl.cpp [hunk 7]
```diff
@@ -24,76 +24,188 @@
...
+            chip::DeviceLayer::Internal::K32WConfigKey(pdm_id_kvs, key_id), (uint8_t *) value, value_size);
+
+        /* save the 'key' in flash such that it can be retrieved later on */
+        if (err == CHIP_NO_ERROR)
+        {
+            ChipLogProgress(DeviceLayer, "KVS, save in flash key [%s] with PDM key: %i", key, key_id + MAX_NO_OF_KEYS);
+            err = chip::DeviceLayer::Internal::K32WConfig::WriteConfigValueStr(
+                chip::DeviceLayer::Internal::K32WConfigKey(pdm_id_kvs, key_id + MAX_NO_OF_KEYS), key, strlen(key));
+        }
+    }
```
Comment (andy31415) [inline src/platform/K32W/KeyValueStoreManagerImpl.cpp:176]: What if false? it sounds like we want to return some error. In this case single exit point makes  the code hard to follow. I see it probably returns CHIP_ERROR_INVALID_ARGUMENT ... but why is that correct?

File: src/platform/K32W/KeyValueStoreManagerImpl.cpp (hunk 5 above)
Comment (andy31415) [reply src/platform/K32W/KeyValueStoreManagerImpl.cpp:51]: you could even do std::bitset and use only 32 bytes of memory.

should static assert that maximum number of keys fits in uint8_t since this is what you use for your set.

File: src/platform/K32W/KeyValueStoreManagerImpl.cpp (hunk 2 above)
Comment (doru91) [reply src/platform/K32W/KeyValueStoreManagerImpl.cpp]: Done, used constexpr, thanks.

File: src/platform/K32W/KeyValueStoreManagerImpl.cpp (hunk 3 above)
Comment (doru91) [reply src/platform/K32W/KeyValueStoreManagerImpl.cpp]: Done, used constexpr and grouped constants in one place.

File: src/platform/K32W/KeyValueStoreManagerImpl.cpp (hunk 4 above)
Comment (doru91) [reply src/platform/K32W/KeyValueStoreManagerImpl.cpp]: Done, used true/false in all the places.

File: src/platform/K32W/KeyValueStoreManagerImpl.cpp (hunk 5 above)
Comment (doru91) [reply src/platform/K32W/KeyValueStoreManagerImpl.cpp]: Replaced list with set and added assert to check that the number of keys is at most kMaxNumberOfKeys.

File: src/platform/K32W/KeyValueStoreManagerImpl.cpp (hunk 6 above)
Comment (doru91) [reply src/platform/K32W/KeyValueStoreManagerImpl.cpp:78]: Thanks, use the proposed logic.

File: src/platform/K32W/KeyValueStoreManagerImpl.cpp (hunk 7 above)
Comment (doru91) [reply src/platform/K32W/KeyValueStoreManagerImpl.cpp:180]: With the newly added assert (assert(g_key_ids_set.size() < kMaxNumberOfKeys)), the condition -> if (put_key) should always be true.

**Comment by doru91** [comment]
@tcarmelveilleux @msandstedt is there any PR opened which requests the change of the KVS API (such that KVS API provides numbers as key identifiers instead of strings)?

**Comment by msandstedt** [comment]
> @tcarmelveilleux @msandstedt is there any PR opened which requests the change of the KVS API (such that KVS API provides numbers as key identifiers instead of strings)?

I am not personally aware of any open PRs.

**Review by andy31415 — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]
Some minor comments inline.

However, the code looks like it is correct and matches its intent.

File: src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp [hunk 8]
```diff
@@ -26,74 +26,190 @@
...
+        }
+        else if (err != CHIP_NO_ERROR)
+        {
+            break;
+        }
+
+        if (key_string_id_size)
+        {
+            g_key_ids_set.insert(key_id);
+            g_kvs_map.insert(std::make_pair(std::string(key_string_id), key_id));
```
[the remark this answers is not in the document]
Comment (doru91) [reply src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp:95]: Could you please give an example flow when the insert operation may fail? Basically, _Put operation makes sure that the key strings are unique and the situation must be the same also when we restore from flash.

[the remark this answers is not in the document]
Comment (msandstedt) [reply src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp:95]: Bugs?  Crashes in the middle of writes?  This is coming from disk.  How do we know what's in there?

It seems a lot easier for the code to audit itself at runtime than it is for developers to analyze all possible corner cases when this code is being written.

[the remark this answers is not in the document]
Comment (doru91) [reply src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp]: Agree, added code that checks the return value from the insert.

File: src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp [hunk 9]
```diff
@@ -26,74 +26,190 @@
...
+        err = chip::DeviceLayer::Internal::K32WConfig::ClearConfigValue(
+            chip::DeviceLayer::Internal::K32WConfigKey(pdm_id_kvs, key_id));
+
+        /* also delete the 'key string' from flash */
+        if (err == CHIP_NO_ERROR)
+        {
+            ChipLogProgress(DeviceLayer, "KVS, delete key [%s] with PDM key: %i", key, key_id + kMaxNumberOfKeys);
+            err = chip::DeviceLayer::Internal::K32WConfig::ClearConfigValue(
+                chip::DeviceLayer::Internal::K32WConfigKey(pdm_id_kvs, key_id + kMaxNumberOfKeys));
+        }
```
[the remark this answers is not in the document]
Comment (doru91) [reply src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp:215]: Improved the logs and also logged failure.

File: src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp [hunk 10]
```diff
@@ -26,74 +26,190 @@
...
+        err = chip::DeviceLayer::Internal::K32WConfig::WriteConfigValueBin(
+            chip::DeviceLayer::Internal::K32WConfigKey(pdm_id_kvs, key_id), (uint8_t *) value, value_size);
+
+        /* save the 'key' in flash such that it can be retrieved later on */
+        if (err == CHIP_NO_ERROR)
+        {
+            ChipLogProgress(DeviceLayer, "KVS, save in flash key [%s] with PDM key: %i", key, key_id + kMaxNumberOfKeys);
+            err = chip::DeviceLayer::Internal::K32WConfig::WriteConfigValueStr(
+                chip::DeviceLayer::Internal::K32WConfigKey(pdm_id_kvs, key_id + kMaxNumberOfKeys), key, strlen(key));
+        }
```
[the remark this answers is not in the document]
Comment (doru91) [reply src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp:183]: Improved the logs and also logged failure.

File: src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp [hunk 11]
```diff
@@ -26,74 +26,190 @@
...
+        {
+            err = CHIP_NO_ERROR;
+            continue;
+        }
+        else if (err != CHIP_NO_ERROR)
+        {
+            break;
+        }
+
+        if (key_string_id_size)
```
[the remark this answers is not in the document]
Comment (doru91) [reply src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp:92]: That's right.

File: src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp [hunk 12]
```diff
@@ -26,74 +26,194 @@
...
+        {
+            g_key_ids_set.insert(key_id);
+            if (!g_kvs_map.insert(std::make_pair(std::string(key_string_id), key_id)).second)
+            {
+                /* key collision is not expected when restoring from flash */
+                abort();
```
Comment (msandstedt) [inline src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp:98]: Sorry to keep hammering on this, but this is actually the worst thing you can do.  This happens at boot.  If you rely on your runtime to clear flash or shepherd firmware updates (likely), you now have a situation where you will be in a never ending reset loop and can never do those things, and thus bad/inconsistent data in flash will have bricked your device.

What you want to do is log this error for starters.  At a bare minimum, this should be visible in dev builds.  Then you can think about what type of remediation or recovery is appropriate.  Some strategies people employ:

* lock out the broken subsystem
* place the entire device in some global 'degraded' state
* self-recovery / cleanup
* (other things)

Guiding principles:

* don't ignore the error
* consider methods of self-recovery
* but do not prevent manual recovery
* if possible, prevent subsequent (incorrect) execution
* BUT do not self-brick

This can quickly get complicated.  If you just want to log the error for now, that's probably fine.

**Review by msandstedt — CHANGES_REQUESTED** [review]
Key collision will likely brick the system with abort().

File: src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp (hunk 12 above)
Comment (doru91) [reply src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp]: @msandstedt thanks a lot for your review, really appreciated. I replaced the abort with logging for the moment. Depending on how the KVS API evolves, I may add a more complex solution to this.

**Review by msandstedt — APPROVED** [review]

**Comment by doru91** [comment]
> @doru91 - this should be ready to merge after the comments from Michael

Solved Michael's comments, the ESP32 build fails independently of the changes from this PR.

File: src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp (hunk 12 above)
Comment (bzbarsky-apple) [reply src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp]: @msandstedt Is this all set in the updated code now?

Comment (msandstedt) [reply src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp]: Yes, current code is logging the error, which is the minimum thing we need and can't trigger the boot-time reset loop.

File: src/platform/K32W/KeyValueStoreManagerImpl.cpp (hunk 7 above)
Comment (msandstedt) [reply src/platform/K32W/KeyValueStoreManagerImpl.cpp]: The assert can:

* get us into a reset loop if existing elements in the persistent store trigger storage of additional elements
* be compiled away and ignored in production code

So this is probably not the greatest solution.

I would think that the right thing to do here is to return CHIP_ERROR_NO_MEMORY, not assert.  It's easy enough to do.

File: src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp [hunk 13]
```diff
@@ -26,74 +26,232 @@
...
+
+    if ((it = g_kvs_map.find(key)) == g_kvs_map.end()) /* new key */
+    {
+        for (key_id = 0; key_id < kMaxNumberOfKeys; key_id++)
+        {
+            std::set<uint8_t>::iterator iter = std::find(g_key_ids_set.begin(), g_key_ids_set.end(), key_id);
+
+            if (iter == g_key_ids_set.end())
+            {
+                assert(g_key_ids_set.size() < kMaxNumberOfKeys);
```
Comment (msandstedt) [inline src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp:165]:
```suggestion
                if (g_key_ids_set.size() >= kMaxNumberOfKeys)
                {
                    return CHIP_ERROR_NO_MEMORY;
                }
```

File: src/platform/K32W/KeyValueStoreManagerImpl.cpp (hunk 7 above)
Comment (msandstedt) [reply src/platform/K32W/KeyValueStoreManagerImpl.cpp]: Added suggestion here:

https://github.com/project-chip/connectedhomeip/pull/9489/files#r744129165

File: src/platform/K32W/KeyValueStoreManagerImpl.cpp (hunk 6 above)
Comment (msandstedt) [reply src/platform/K32W/KeyValueStoreManagerImpl.cpp]: > Either consistently clear the error or return error on first failure. It sounds like the logic should be:
> 
>     * if NOT_FOUND then set error to NO_ERROR and continue
> 
>     * on any error except NO_ERROR or NOT_FOUND, return an error

Current code sets g_restored_from_flash when returning the error.

I think this is not correct.  This means that subsequent ops will silently ignore the fact that initialization has actually failed.

File: src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp [hunk 14]
```diff
@@ -26,74 +26,232 @@
...
+        if (err == CHIP_DEVICE_ERROR_CONFIG_NOT_FOUND)
+        {
+            err = CHIP_NO_ERROR;
+            continue;
+        }
+        else if (err != CHIP_NO_ERROR)
+        {
+            ChipLogProgress(DeviceLayer, "KVS, Error while restoring Matter key [%s] from flash with PDM id: %i", key_string_id,
+                            pdm_internal_id);
+            break;
```
Comment (msandstedt) [inline src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp:92]:
```suggestion
            return err;
```

Comment (msandstedt) [reply src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp:92]: Else we set g_restored_from_flash=true, even when init failed.

File: src/platform/K32W/KeyValueStoreManagerImpl.cpp (hunk 6 above)
Comment (msandstedt) [reply src/platform/K32W/KeyValueStoreManagerImpl.cpp]: Added suggestion here: https://github.com/project-chip/connectedhomeip/pull/9489/files#r744130070.

File: src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp (hunk 13 above)
Comment (msandstedt) [reply src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp:165]: Else we can get in a failure / reset loop, or (if assert is compiled out of the build) ignore the persistent storage bounds.

Assert shouldn't really be used for things that have a dependency on things from disk.

File: src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp (hunk 14 above)
Comment (doru91) [reply src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp]: Right, solved.
